---
title: "CIT analysis"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/mmQTL
# rm -rf combined_analysis/eval_resid_in_each_cache/
ml git pandoc R/4.0.2
git pull
R

system("git pull")


rmarkdown::render("cit_analysis.Rmd")



--->

```{r setup, include=FALSE}
suppressPackageStartupMessages({
library(data.table)
library(ggplot2)
library(cowplot)
library(vcd)
library(GenomicRanges)
library(AnnotationHub)
library(ensembldb)
library(cit)
library(synapser)
})

synLogin()

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```


```{r functions}
# For ENSEMBL id ENSG00000279457.4, return ENSG00000279457
trim_ensembl_ids = function(x){
  gsub("(.*)\\.(.*)", "\\1", x) 
}

# Add column Symbol using column gene_id storing ENSEMBL id
getGeneSymbol = function( df, column="row.names"){

  if( ! is(df, "data.frame") ){
    df = as.data.frame(df)
  }

  # Load ENSEMBL v96 database
  # see https://www.bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/ensembldb.html#10_getting_or_building_ensdb_databasespackages
  ah <- AnnotationHub()
  ensdb = ah[["AH69187"]] # ENSEMBL v96

  if( column == "row.names"){    
    geneID = rownames(df)
  }else{
    geneID = df[[column]]
  }

  ensIDs = trim_ensembl_ids( geneID )

  geneSymbol = mapIds(ensdb, keys=ensIDs, keytype="GENEID", column="GENENAME")
  df$Symbol = geneSymbol

  entrez = mapIds(ensdb, keys=ensIDs, keytype="GENEID", column="ENTREZID")
  df$Entrez = entrez

  df_info = ensembldb::select(ensdb, keys=ensIDs, keytype="GENEID", column=c("SEQNAME", "GENESEQSTART", "GENESEQEND"))

  # df2 = merge( data.frame(GENEID = ensIDs, stringsAsFactors=FALSE), chroms, by='GENEID')

  idx = match(ensIDs, df_info$GENEID)
  # df$Chrom = c()
  df$Chrom = df_info$SEQNAME[idx]
  # df$Start = c()
  df$Start = df_info$GENESEQSTART[idx]
  # df$End = c()
  df$End = df_info$GENESEQEND[idx]

  df
}
```

```{r read.xQTL}
# caQTL fine-mapping results # old syn24201357
df_caqtl = fread( synGet('syn24911652')$path, header=FALSE)
colnames(df_caqtl) = c("Chr", "Peak", "eQTL_order", "Variant", "PIP")
df_caqtl[,Chr := c()]
setkey(df_caqtl, 'Variant')
df_caqtl_best = df_caqtl[,.SD[which.max(PIP),],by="Variant"]

# peak location
df_peak = fread( synGet('syn24201356')$path )
colnames(df_peak) = c("Chr", "start", "end", 'Peak')
gr_peak = with(df_peak, GRanges( Chr, IRanges(start, end, name=Peak)))


# eQTL fine-mapping results
df_eqtl = fread( synGet('syn24178479')$path )
colnames(df_eqtl)[5] = "PIP"
df_eqtl[,Chr := c()]
setkey(df_eqtl, 'Variant')
df_eqtl_best = df_eqtl[,.SD[which.max(PIP),],by="Variant"]

# location of genes
df_gene = getGeneSymbol( df_eqtl[,data.frame(Gene = unique(Gene))], "Gene")
gr_gene = with(df_gene, GRanges(Chrom, IRanges(Start, End, name=Gene)))

# ABC links
df_abc = readRDS( synGet('syn24346713')$path )$ALL_CPM1
# df_abc = df_abc[df_abc$ABC.Score > .1,]

# Merge eQTL and caQTL
# only merge when both a gene and peak have a candidate causal variant
df_merge = merge(df_eqtl, df_caqtl_best, by="Variant")
df_merge[is.na(PIP.x), PIP.x:=0]
df_merge[is.na(PIP.y), PIP.y:=0]
df_merge[,eQTL_order.x:=c()]
df_merge[,eQTL_order.y:=c()]
df_merge[,combo := paste(Peak, Gene, sep='_') ]
```

```{r read.data}

chromAccess = readRDS("/sc/arion/projects/Microglia/atacseq/step_qc_analyses/Model1_ATACseq_norm__QuantPerMig_BIC_4in0.05_CPM_1in0.2_BBrandom_atFDR_0.05/files/KeepDxResidualized_matrix.RDs")

geneExpr  <- readRDS("/sc/arion/projects/Microglia/rnaseq/step_qc_analyses//NotStrict_SampQC_Model1_gene_norm__QuantPerMig_BIC_4in0.05_CPM_1in0.2_BBrandom_atFDR_0.05/KeepDxResidualized_matrix.RDs")

genotypeVCF = "/sc/arion/projects/Microglia/genotyping/Combined_Phase1and2/Merged_Chr1toX_FixedNames.vcf.gz"
```












