---
title: "Designing MPRA from BREMA"
author: "[Gabriel Hoffman](http://gabrielhoffman.github.io)"
date: "Run on `r Sys.time()`"
output: html_document
---


<!---

cd /Users/gabrielhoffman/workspace/repos/mmQTL_plots
# rm -rf combined_analysis/combine_all_EPHA1_cache

cd /hpc/users/hoffmg01/work/mmQTL
ml git
git pull
R

rmarkdown::render("design_MPRA.Rmd")

# scp design_MPRA.html sklar1:/hpc/users/hoffmg01/www/software/MPRA


--->
 


 ```{r, eval=FALSE, echo=FALSE}
library(synapser)
library(knit2synapse)
synLogin()

# synDelete('syn25613257')


# Run code and upload results to Synapse
knit2synapse::createAndKnitToFolderEntity(file = "design_MPRA.Rmd",
                                          parentId ="syn25585302",
                                          folderName = 'Design of MPRA',
                                          overwrite=TRUE)



```

```{r setup, include=FALSE, echo=FALSE}
library(data.table)
library(ggplot2)
library(viridis)
library(synapser)
library(knit2synapse)
synLogin()

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  dev = c('png', 'pdf'),
  cache = TRUE,
  cache.lazy = FALSE)
```

df1 = fread(synGet('syn25961123')$path)
df2 = fread(synGet('syn25961118')$path)


```{r synapse.parameters, include=FALSE, cache=TRUE}
library(githubr)
parentId = 'syn25585302';
activityName = 'Design of MPRA';
activityDescription = 'Design of MPRA';
thisFileName <- 'design_MPRA.Rmd'
# Github link
thisRepo <- getRepo(repository = "GabrielHoffman/mmQTL", ref="branch", refName='main')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=thisFileName)

ALL_USED_IDs = c()
```


```{r read.data, message=FALSE}
# fine-mapping results
df_in = fread(synGet('syn25592273')$path) # V2
ALL_USED_IDs = c(ALL_USED_IDs, 'syn25592273')

# df_eQTL = fread(synGet('syn25592272')$path) # V2
# ALL_USED_IDs = c(ALL_USED_IDs, 'syn25592272')

# variant positions
df_positions = fread(synGet('syn25955442')$path) # V2
ALL_USED_IDs = c(ALL_USED_IDs, 'syn25592274')
colnames(df_positions) = c("Variant", "chrom", "position_hg19", "REF", "ALT")

df = merge(df_in, df_positions, by='Variant')
setkey(df, "Variant")
```

```{r load.eqtl.info}
# get SNPs with large p-values
# z_score_fixed > 0 means Assessed_allele is more highly expressed
df_eqtl_all = fread(synGet('syn25961123')$path, select = c('Gene', 'Variant', 'eQTL_order', 'p_fixed', 'p_random', 'z_score_fixed', 'Non_assessed_allele', 'Assessed_allele'))
setkey(df_eqtl_all, "Variant")

# merge with position
# only keep SNPs
df_eqtl_all = merge(df_eqtl_all, df_positions[nchar(REF) == nchar(ALT),], by="Variant")

# If z_score_fixed > 0, then Assessed_allele is more highly expressed
df_eqtl_all[,HighAllele := ifelse(z_score_fixed > 0, Assessed_allele, Non_assessed_allele) ]

# get MAF for each variant
df_maf = fread(synGet('syn25953082')$path)

df_eqtl_all = merge(df_eqtl_all, df_maf[,c('SNP', 'MAF')], by.x="Variant", by.y="SNP")

# merge with HighAllele and MAF
df = merge(df, df_eqtl_all[,c("Variant", "Gene", "eQTL_order", "HighAllele", "MAF")], by=c("Variant", "Gene", "eQTL_order"))
```


```{r ATACseq, message=FALSE}
library(GenomicRanges)

# Create snATAC-seq data liftOver to hg19
gr.lst = readRDS( synGet('syn25612085')$path)

# Get GenomicRanges from eQTL variants
gr.eQTL = with(df, GRanges(Chr, IRanges(position_hg19, position_hg19, name=Variant)))

# add columns indicating presence in ATAC peaks
for(cellType in names(gr.lst) ){

  df[[paste0("ATAC_",cellType)]] = overlapsAny(gr.eQTL, gr.lst[[cellType]])
}
```

# Count variants
```{r count.variants, fig.width=5, fig.height=5}

df.count = lapply(1:5, function(ord){
  df_filter = df[(eQTL_order == ord) & (PP > 0.1),]

  x = seq(min(df_filter$PP), 1, length=1000)

  y = sapply(x, function(i){
  	sum(df_filter$PP > i)
  	})

   data.frame(order=as.character(ord), x,y)
})
df.count = do.call(rbind, df.count)
df.count$order = factor(df.count$order, as.character(1:10))

ggplot(df.count, aes(x,y, color=order)) + geom_line() + theme_bw() + xlab("Posterior inclusion probability") + ylab("Number of variants") + ggtitle("Fine-mapping: by variant") + xlim(0, 1) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_brewer(palette="Set1", name="eQTL order")
```


# Count genes
```{r count.genes, fig.width=5, fig.height=5}

df.count = lapply(1:5, function(ord){

  df2 = df[(eQTL_order == ord),.SD[which.max(PP)], by="Gene"]
  df2 = df2[order(PP),]

  x = seq(0, 1, length=1000)

  y = sapply(x, function(i){
  	sum(df2$PP > i)
  })
  data.frame(order=as.character(ord), x,y)
})
df.count = do.call(rbind, df.count)
df.count$order = factor(df.count$order, as.character(1:5))

ggplot(df.count, aes(x,y, color=order)) + geom_line() + theme_bw() + xlab("Posterior inclusion probability") + ylab("Number of genes") + ggtitle("Fine-mapping: by gene") + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_brewer(palette="Set1", name="eQTL order")
```

# Select variants
## PP > 0.1
```{r filter, fig.width=12, fig.height=5}
df_one = df[eQTL_order==1,]

par(mfrow=c(1,3))
hist(df_one[PP > .1,PP], xlab="Posterior probability")
hist(df_one[PP > .1,Dis_to_TSS], xlab="Distance to TSS")
hist(log10(abs(df_one[PP > .1,Dis_to_TSS])), xlab="Distance to TSS (log10)")
```

Number of variants: `r nrow(df_one[PP > .1,])`

Number of genes: `r df_one[PP > .1,length(unique(Gene))]`


## PP > 0.01, sample 10K variants in proportion to PP
```{r filter2, fig.width=12, fig.height=5}
df_one = df[(eQTL_order==1)&(PP > 0.01),]

set.seed(1)
idx = sample.int(nrow(df_one), 20000, prob=df_one$PP)

par(mfrow=c(1,3))
hist(df_one[idx,PP], xlab="Posterior probability")
hist(df_one[idx,Dis_to_TSS], xlab="Distance to TSS")
hist(log10(abs(df_one[idx,Dis_to_TSS])), xlab="Distance to TSS (log10)")
```

Number of variants with PP>0.1: `r nrow(df_one[idx,][PP > .1,])`

Number of genes: `r df_one[idx,][PP > .1,length(unique(Gene))]`

# Evaluate retention rate at increasing number of variants
Here, variants are selected in proprotion to their posterior probability.

With > 20-30K variants, most of the high-scoring variants are included, and the remaining available MPRA sequences are devoted to low-scoring variants.

```{r test, fig.width=5, fig.width=5}
set.seed(1)
n_array = c(1:6)*10000
idx.lst = lapply( n_array, function(n){
  sort(sample.int(nrow(df_one), n, prob=df_one$PP))
})
names(idx.lst) = n_array

df_stats = lapply( n_array, function(n){

  idx = idx.lst[[as.character(n)]] 

  xvalue = seq(0.01, 1, length.out=100)
  df_stats = lapply(xvalue, function(x){
    # fraction of variance that pass cutoff
    exceed = nrow(df_one[idx,][PP > x,])
    frac = exceed / nrow(df_one[PP > x,])

    # fraction of variants in ATAC-seq peaks
    cols = grep("^ATAC_", colnames(df_one))
    counts = apply(df_one[idx,][PP > x,][,..cols], 2, sum)
    data.frame(n, x, frac, exceed, t(counts))
  })
  do.call(rbind, df_stats) 
})
df_stats = do.call(rbind, df_stats)

ggplot(df_stats, aes(x, frac, color=factor(n))) + geom_line() + theme_classic() + xlab('Posterior probability cutoff') + ylab("Fraction of variants retained") + xlim(0, 1) + ylim(0, 1) + scale_color_viridis_d("# of variants") + theme(aspect.ratio=1)
```

For each cell type in the snATAC-seq from [Corces, et al. 2020](https://www.nature.com/articles/s41588-020-00721-x) and [GSE147672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE147672), the fraction of variants in ATAC-seq peaks is reported across posterior probability cutoffs.  With fewer variants in the MPRA, more high scoring variants are selected, and these are more likely to be in ATAC peaks.

```{r ATAC.integrate, fig.width=9, fig.height=9}

df_melt = reshape2::melt(df_stats, id.vars=c('n', 'x', "exceed", "frac"))

ggplot(df_melt, aes(x, value/exceed, color=factor(n))) + geom_line() + theme_classic() + facet_wrap(~variable, ncol=3) + xlab('Posterior probability cutoff') + ylab("Fraction of variants exceeding PIP cutoff in ATAC peaks") + scale_color_viridis_d(name = "# of variants") + theme(aspect.ratio=1)
```

Here, the fration of *total* variants is show.  

```{r ATAC.integrate2, fig.width=9, fig.height=9}

df_melt = reshape2::melt(df_stats, id.vars=c('n', 'x', "exceed", "frac"))

ggplot(df_melt, aes(x, value/n, color=factor(n))) + geom_line() + theme_classic() + facet_wrap(~variable, ncol=3) + xlab('Posterior probability cutoff') + ylab("Fraction of total variants in ATAC peaks") + scale_color_viridis_d(name = "# of variants") + theme(aspect.ratio=1)
```

Show distance to TSS stratified by posterior probability

```{r ATAC.integrate3, fig.width=9, fig.height=5}
df_stats = lapply( names(idx.lst), function(n){

  idx = idx.lst[[as.character(n)]] 

  xvalue = seq(0.01, .99, length.out=100)
  df_stats = lapply(xvalue, function(x){
    data.frame(n, x, dTSS = df_one[idx,][PP > x,Dis_to_TSS])
  })
  do.call(rbind, df_stats)
})
df_stats = do.call(rbind, df_stats)

ggplot(df_stats, aes(abs(dTSS), group=x, color=x)) + geom_density() + scale_x_log10() + scale_color_viridis_c("Posterior probability", limits=c(0,1)) + theme_classic() + theme(aspect.ratio=1) + xlab("Distance to TSS") + facet_wrap(~n) 
```





```{r storeResults}
# Code
CODE <- Folder(name = "Design of MPRA", parentId = 'syn25585302')
CODE <- synStore(CODE)

sapply( names(idx.lst), function(n){

  file = paste0('MPRA_BREMA_variants_', n, '.tsv')

  idx = idx.lst[[as.character(n)]] 
  
  write.table(df_one[idx,], file = file, sep = '\t', row.names=FALSE, quote=FALSE)
  file.gz = R.utils::gzip(file, overwrite=TRUE)

  # Upload file
  COV_OBJ = synapser::File(file.gz, name = file.gz, parentId = CODE$properties$id)
  COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                     executed = thisFile, activityDescription = activityDescription)
})
```





# Select negative variants
Common variants with high p-value (>0.3), high MAF (>0.2), *in* ATAC-seq peaks
```{r load.filter.eqtl}

# Select variants with large p-values and MAF
df_eqtl_subset = df_eqtl_all[MAF > 0.2,.SD[which.min(p_fixed),],by="Variant"]
df_eqtl_subset2 = df_eqtl_subset[(p_fixed > .1)&(p_random>.1),]
```


```{r Additional.filtering}
# convert to GRanges
gr_eqtl = with(df_eqtl_subset2, GRanges(paste0('chr', chrom), IRanges(position_hg19, position_hg19, name=Variant   )))
gr_eqtl = unique(gr_eqtl)

# identify SNPs in ATAC-seq peaks for each cell type
res = lapply(names(gr.lst), function(cellType){

  idx = overlapsAny( gr_eqtl, gr.lst[[cellType]])

  data.table(Variant = names(gr_eqtl)[idx], Count = 1)
})
res = do.call(rbind, res)

# count the number of cell types that have ATAC peaks over each SNP
df_ATAC_negSNP = res[,data.table(Count_ATAC = sum(Count)), by="Variant"]

# hist(df_ATAC_negSNP$Count_ATAC)

df_cat1 = merge(df_ATAC_negSNP[Count_ATAC > 2,], df_positions, by="Variant")
df_cat1 = merge(df_cat1, df_maf[,c("SNP", "MAF")], by.x="Variant", by.y="SNP")

file = "negative_control_variants_Type_1.tsv"
write.table(df_cat1, file = file, sep = '\t', row.names=FALSE, quote=FALSE)
file.gz = R.utils::gzip(file, overwrite=TRUE)

# get Synapse info
CODE <- Folder(name = "Design of MPRA", parentId = 'syn25585302')
CODE <- synStore(CODE)

# Upload file
COV_OBJ = synapser::File(file.gz, name = file.gz, parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                 executed = thisFile, activityDescription = activityDescription)
```



Within locus with high PP, select nearby variants that are not in candidate causal set.  There variants are 1) 300-1000 bp from the candidate causal variant with PIP > 0.99, 2) have p < 1e-8, 3) have PIP = 0.
```{r near.highPP}
# select SNPs with high PIP
df_high = df_one[PP > .98,]

# location of candidate causal variants
gr_high = with(df_high, GRanges(Chr, IRanges(position_hg19, position_hg19, name=Variant)))

idx_gene = which(df_high$Gene %in% df_eqtl_all[,unique(Gene)])

setkey(df_eqtl_all, "Gene")

# just primary eQTL
df_eqtl_1 = df_eqtl_all[(eQTL_order==1),]
setkey(df_eqtl_1, "Gene")

df_cat2 = lapply( idx_gene, function(i){
  cat("\r", i, "   ")

  # For each strong candidate causal SNP, select another SNP that is 
  # nearby, correlated, small p-value, in primary signal, but not estimated to be causal
  ensGene = df_high[i,Gene]
  SNP = df_high[i,Variant]

  # get eQTL results for this gene
  df = df_eqtl_1[Gene == ensGene,][order(position_hg19),]

  # convert to GenomicRanges
  gr = with(df, GRanges(paste0('chr', chrom), IRanges(position_hg19, position_hg19, name=Variant, score=p_fixed)))

  # compute distance to target SNP
  d = distance(gr_high[i], gr)
  idx = which((d > 300) &(d <1200 ))

  # get statistics for tnear by SNPs
  # drop variants that have non-zero PIP's
  df = data.table(Variant = names(gr[idx]), p.value = score(gr[idx]), distance = d[idx], target=SNP)
  df = merge(df, df_in, by="Variant", all.x=TRUE)
  df = df[order(p.value),][is.na(PP),]
  df[(p.value < 1e-7),1:4]
})
df_cat2 = do.call(rbind, df_cat2)

df_cat2 = merge(df_cat2, df_positions, by="Variant")
df_cat2 = merge(df_cat2, df_maf[,c("SNP", "MAF")], by.x="Variant", by.y="SNP")

file = "negative_control_variants_Type_2.tsv"
write.table(df_cat2, file = file, sep = '\t', row.names=FALSE, quote=FALSE)
file.gz = R.utils::gzip(file, overwrite=TRUE)

# Upload file
COV_OBJ = synapser::File(file.gz, name = file.gz, parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                 executed = thisFile, activityDescription = activityDescription)
```


### Source
[markdown](`r thisFile`)





